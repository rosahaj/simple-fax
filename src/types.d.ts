export type EnvelopeFiletype = 'PS' | 'PDF'

export interface EnvelopeOptions {
  /**
   * Email address
   */
  username: string
  /**
   * Password
   */
  password: string
  /**
   * Recipient fax number
   */
  faxnumber: string
  /**
   * Base64-encoded PDF or PS document
   */
  data: string
  /**
   * Either "PDF" or "PS"
   */
  filetype: EnvelopeFiletype
  /**
   * Sender information (included in fax header)
   */
  identifier: string
  /**
   * Callback URL
   *
   * After transmission, a `GET` request with the following parameters is sent to the callback URL:
   * - `faxid`
   * - `status` (either "done" or "failed")
   * - `reason` (error message)
   */
  statusurl: string
}
