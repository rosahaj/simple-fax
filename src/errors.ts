// export type ErrorCode = 'INVALID_CREDENTIALS' | 'INVALID_FAX_NUMBER' | 'INSUFFICIENT_BALANCE' | 'INVALID_DOCUMENT' | 'INTERNAL_ERROR'
export enum ErrorCode {
  InvalidCredentials = 'INVALID_CREDENTIALS',
  InvalidFaxNumber = 'INVALID_FAX_NUMBER',
  InsufficientBalance = 'INSUFFICIENT_BALANCE',
  FaultyDocument = 'FAULTY_DOCUMENT',
  InternalError = 'INTERNAL_ERROR',
  Other = 'OTHER',
}

export class SimpleFaxError extends Error {
  code: ErrorCode

  constructor(msg: string) {
    super(msg)

    switch (msg) {
      case 'Zugangsdaten falsch':
        this.code = ErrorCode.InvalidCredentials
        break

      case 'Keine g√ºltige Faxnummer':
        this.code = ErrorCode.InvalidFaxNumber
        break

      case 'Guthaben reicht nicht aus':
        this.code = ErrorCode.InsufficientBalance
        break

      case 'Dokument fehlerhaft':
        this.code = ErrorCode.FaultyDocument
        break

      default:
        this.code = ErrorCode.Other
        break
    }

    Object.setPrototypeOf(this, SimpleFaxError.prototype)
  }
}
