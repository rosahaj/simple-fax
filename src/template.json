{
  "elements": [
    {
      "type": "element",
      "name": "soapenv:Envelope",
      "attributes": {
        "xmlns:xsi": "http://www.w3.org/2001/XMLSchema-instance",
        "xmlns:xsd": "http://www.w3.org/2001/XMLSchema",
        "xmlns:soapenv": "http://schemas.xmlsoap.org/soap/envelope/",
        "xmlns:urn": "urn:soapservice"
      },
      "elements": [
        { "type": "element", "name": "soapenv:Header" },
        {
          "type": "element",
          "name": "soapenv:Body",
          "elements": [
            {
              "type": "element",
              "name": "urn:sendfax",
              "attributes": {
                "soapenv:encodingStyle": "http://schemas.xmlsoap.org/soap/encoding/"
              },
              "elements": [
                {
                  "type": "element",
                  "name": "username",
                  "attributes": { "xsi:type": "xsd:string" },
                  "elements": [{ "type": "text", "text": "" }]
                },
                {
                  "type": "element",
                  "name": "password",
                  "attributes": { "xsi:type": "xsd:string" },
                  "elements": [{ "type": "text", "text": "" }]
                },
                {
                  "type": "element",
                  "name": "faxnumber",
                  "attributes": { "xsi:type": "xsd:string" },
                  "elements": [{ "type": "text", "text": "" }]
                },
                {
                  "type": "element",
                  "name": "data",
                  "attributes": { "xsi:type": "xsd:string" },
                  "elements": [{ "type": "text", "text": "" }]
                },
                {
                  "type": "element",
                  "name": "filetype",
                  "attributes": { "xsi:type": "xsd:string" },
                  "elements": [{ "type": "text", "text": "" }]
                },
                {
                  "type": "element",
                  "name": "identifier",
                  "attributes": { "xsi:type": "xsd:string" },
                  "elements": [{ "type": "text", "text": "" }]
                },
                {
                  "type": "element",
                  "name": "statusurl",
                  "attributes": { "xsi:type": "xsd:string" },
                  "elements": [{ "type": "text", "text": "" }]
                }
              ]
            }
          ]
        }
      ]
    }
  ]
}
