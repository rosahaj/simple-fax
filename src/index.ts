import { js2xml } from 'xml-js'
import type { Options } from 'ky'
import ky from 'ky'
import { SimpleFaxError } from './errors.js'
import type { EnvelopeOptions } from './types.js'
import template from './template.json'

/**
 * Creates XML string to be used as request body
 */
export function createEnvelope(options: EnvelopeOptions): string {
  const envelope = structuredClone(template)
  const elements = envelope.elements[0].elements[1].elements![0].elements

  elements.forEach(
    element =>
      (element.elements[0].text = (
        options as EnvelopeOptions & { [key: string]: string }
      )[element.name]),
  )

  return js2xml(envelope)
}

/**
 * Sends a fax using the simple-fax.de SOAP API
 *
 * Note: `data` needs to be a base64-encoded string
 *
 * Throws a `SimpleFaxError` if the simple-fax.de server responds with an error
 *
 * @param options API options
 * @param kyOptions Additional request options passed to ky
 * @returns FaxId
 */
export async function sendFax(
  options: EnvelopeOptions,
  kyOptions?: Options,
): Promise<string> {
  const envelope = createEnvelope(options)

  const response = await ky
    .post('http://longisland.simple-fax.de/soap/index.php', {
      ...kyOptions,
      body: envelope,
      headers: {
        ...kyOptions?.headers,
        'Content-Type': 'text/xml;charset=UTF-8',
        'SOAPAction': `"urn:soapservice#Service#sendfax"`,
      },
    })
    .text()

  // Response format: "status;message."
  const [status, message] = response.trim().replace(/\.$/, '').split(';')

  if (status === 'success')
    return message
  else throw new SimpleFaxError(message)
}
