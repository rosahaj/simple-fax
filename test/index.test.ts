import { describe, expect, it } from 'vitest'
import { createEnvelope } from '../src'
import type { EnvelopeOptions } from '../src/types'

describe('envelope', () => {
  it('is valid', () => {
    const options: EnvelopeOptions = {
      username: 'username',
      password: 'password',
      faxnumber: 'faxnumber',
      data: 'data',
      filetype: 'PDF',
      identifier: 'identifier',
      statusurl: 'statusurl',
    }

    const xml = createEnvelope(options)

    expect(xml).toMatchSnapshot()

    expect(xml).toContain(
      `<username xsi:type="xsd:string">${options.username}</username>`,
    )
    expect(xml).toContain(
      `<password xsi:type="xsd:string">${options.password}</password>`,
    )
    expect(xml).toContain(
      `<faxnumber xsi:type="xsd:string">${options.faxnumber}</faxnumber>`,
    )
    expect(xml).toContain(`<data xsi:type="xsd:string">${options.data}</data>`)
    expect(xml).toContain(
      `<filetype xsi:type="xsd:string">${options.filetype}</filetype>`,
    )
    expect(xml).toContain(
      `<statusurl xsi:type="xsd:string">${options.statusurl}</statusurl>`,
    )
  })
})
