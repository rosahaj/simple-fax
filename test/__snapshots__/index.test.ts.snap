// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`envelope > is valid 1`] = `"<soapenv:Envelope xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:xsd="http://www.w3.org/2001/XMLSchema" xmlns:soapenv="http://schemas.xmlsoap.org/soap/envelope/" xmlns:urn="urn:soapservice"><soapenv:Header/><soapenv:Body><urn:sendfax soapenv:encodingStyle="http://schemas.xmlsoap.org/soap/encoding/"><username xsi:type="xsd:string">username</username><password xsi:type="xsd:string">password</password><faxnumber xsi:type="xsd:string">faxnumber</faxnumber><data xsi:type="xsd:string">data</data><filetype xsi:type="xsd:string">PDF</filetype><identifier xsi:type="xsd:string">identifier</identifier><statusurl xsi:type="xsd:string">statusurl</statusurl></urn:sendfax></soapenv:Body></soapenv:Envelope>"`;
